### PR TITLE
fix(path) Support absolute path with spaces or special characters

### PR DESCRIPTION
**Pull request by rbioteau**
Absolute to relative path replacement was broken when the absolute file
path was conataing space characters. Space characters are subsitute with
their percent-encoding value in html `` src attribute making the
regular expression pattern invalid in this case and resulting with a
preview having broken image references when the path to temp folder
contains spaces (project name or OS account name).

I don't known if those path can be at other places in the html document
where the path is not encoded. If this is the case we might need 2
regular pattern expression to replace.

In addition I have quoted the regexp pattern to support reserved regexp
characters in the path like `()` or `$`

- [x] Validated on Windows
- [x] Validated on Linux

Relates to https://github.com/de-jcup/eclipse-asciidoctor-editor/issues/134, https://github.com/de-jcup/eclipse-asciidoctor-editor/issues/268, https://github.com/de-jcup/asp/issues/23

### CODE REVIEW

File: asciidoctor-editor-plugin/src/test/java/de/jcup/asciidoctoreditor/preview/AsciiDoctorEditorBuildSupportTest.java [hunk 1]
```diff
@@ -0,0 +1,70 @@
+/*
```
Comment (de-jcup) [inline asciidoctor-editor-plugin/src/test/java/de/jcup/asciidoctoreditor/preview/AsciiDoctorEditorBuildSupportTest.java]: Nice test, but please move it into src/test/java-eclipse - reaon: It works inside eclipse, but gradle build does not have all eclipse dependencies. So gradle build do only inspect /src/test/java

File: asciidoctor-editor-plugin/src/main/java-eclipse/de/jcup/asciidoctoreditor/preview/AsciiDoctorEditorBuildSupport.java [hunk 2]
```diff
@@ -324,10 +325,14 @@ protected boolean isNeedingAHiddenEditorFile(File editorFileOrNull, File fileToC
 
     protected Pattern createRemoveAbsolutePathToTempFolderPattern() {
         Path tempFolder = getEditor().getWrapper().getTempFolder();
-        String absolutePathToTempFolder = tempFolder.toFile().getAbsolutePath();
```
Comment (de-jcup) [inline asciidoctor-editor-plugin/src/main/java-eclipse/de/jcup/asciidoctoreditor/preview/AsciiDoctorEditorBuildSupport.java:327]: Interesting: 
At my linux system I used an images directory and created a test asciidoc file containing

```asciidoc
image::images folder with spaces/xasciidoctor-editor-logo.png[]
image::images folder with spaces/xdot-example.svg[]
```
Defined image content ala: 
image

When using your changes I see
image


when using my origin codeScreenshot at 2019-11-16 01-18-57, I see
image

So i had no problems with pathes containing spaces (no matter if inside path to asciidocfile nor to images )
I only had problems after your changes.

Do you have use a windows machine? I tried out with Linux. Or did you used no images directory?

Please do me a favor and add a similar setup like mine to 'eclipse-asciidoctor-editor/asciidoctor-editor-other/testscripts/images' and a test asciidoc file inside 
`eclipse-asciidoctor-editor/asciidoctor-editor-other/testscripts/bugfixes`
so we have a simple test case.

File: asciidoctor-editor-plugin/src/main/java-eclipse/de/jcup/asciidoctoreditor/preview/AsciiDoctorEditorBuildSupport.java [hunk 3]
```diff
@@ -324,10 +325,14 @@ protected boolean isNeedingAHiddenEditorFile(File editorFileOrNull, File fileToC
...
-        String asciidocOutputAbsolutePath = absolutePathToTempFolder.replace('\\', '/');
-        asciidocOutputAbsolutePath += "/"; // relative path is without leading /
-        return Pattern.compile(asciidocOutputAbsolutePath);
+        //Convert to URI as asciidoc convert file path to URI in html document.
+        //So if the path contains a space or a special character it will be percent encoded
+        URI absolutePathToTempFolder = tempFolder.toFile().toURI();
+        String path = absolutePathToTempFolder.getRawPath();
+        if (path.startsWith("/")) {
```
Comment (rbioteau) [inline asciidoctor-editor-plugin/src/main/java-eclipse/de/jcup/asciidoctoreditor/preview/AsciiDoctorEditorBuildSupport.java]: I'm on windows and there might be a system compatibility issue with this starting `/`. I have to remove it on windows, maybe not on Unix system, can you try to leave it and make a test ?

File: asciidoctor-editor-plugin/src/main/java-eclipse/de/jcup/asciidoctoreditor/preview/AsciiDoctorEditorBuildSupport.java (hunk 2 above)
Comment (rbioteau) [reply asciidoctor-editor-plugin/src/main/java-eclipse/de/jcup/asciidoctoreditor/preview/AsciiDoctorEditorBuildSupport.java:327]: The issue is not about the relative path in the adoc file. But the absolute path used when generating the preview. To reproduce you have to use spaces in the name of the parent project of the adoc file.

I'll have a look to the test setup

File: asciidoctor-editor-plugin/src/test/java/de/jcup/asciidoctoreditor/preview/AsciiDoctorEditorBuildSupportTest.java (hunk 1 above)
Comment (rbioteau) [reply asciidoctor-editor-plugin/src/test/java/de/jcup/asciidoctoreditor/preview/AsciiDoctorEditorBuildSupportTest.java]: I have moved the test and added the Mockito dependency in the plugin MANIFEST (as optional).
Tell me if it is ok for you and if you can reproduce the problem with the test.

File: asciidoctor-editor-plugin/src/main/java-eclipse/de/jcup/asciidoctoreditor/preview/AsciiDoctorEditorBuildSupport.java (hunk 3 above)
Comment (rbioteau) [reply asciidoctor-editor-plugin/src/main/java-eclipse/de/jcup/asciidoctoreditor/preview/AsciiDoctorEditorBuildSupport.java]: I tried on Linux and there is a different path behavior between Linux and windows.
I have update the PR to take this difference into account.

**Pull request merged**